src/libraries/System.Security.Cryptography.Cose/src/System/Security/Cryptography/Cose/ToBeSignedBuilder.cs

```diff
@@ -0,0 +1,43 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Collections.Generic;
+using System.Text;
+
+namespace System.Security.Cryptography.Cose
+{
+    internal abstract class ToBeSignedBuilder : IDisposable
+    {
+        // arg is passthrough - we don't do anything with it but all usages need to pass in extra Span and it's not allowed to do through closure.
+        internal delegate void ToBeSignedOperation(Span<byte> arg, ReadOnlySpan<byte> data);
+
+        private bool _disposed;
+
+        internal static ToBeSignedBuilder Create(KeyType keyType, HashAlgorithmName hashAlgorithmName)
+            => keyType switch
+            {
+                KeyType.MLDsa => new PureDataToBeSignedBuilder(),
+                _ => new HashToBeSignedBuilder(hashAlgorithmName),
+            };
+
+        public void Dispose()
+        {
+            if (!_disposed)
+            {
+                Dispose(true);
+                _disposed = true;
+                GC.SuppressFinalize(this);
```

Comment:
Since the type hierarchy is `internal` and we know we won't use finalizers, we can use DisposeCore instead of Dispose(bool), if preferred.